tests/e2e/structural/BUILD.bazel

```diff
@@ -0,0 +1,49 @@
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Tests for different generic program structures, including things like:
+#   * [float32, int32, etc.] data type support
+#   * multiple functions in a single module
+#   * functions with [0, 1, 2+] inputs
+#   * functions with [0, 1, 2+] outputs
+
+load("//build_tools/bazel:enforce_glob.bzl", "enforce_glob")
+load("//build_tools/bazel:iree_check_test.bzl", "iree_check_single_backend_test_suite")
+
+package(
+    features = ["layering_check"],
+    licenses = ["notice"],  # Apache 2.0
+)
+
+# Ideally we would use 'check' tests rather than 'lit' tests for as many tests
+# as possible, to allow for compilation on host machines and test execution on
+# other platforms (with build system support for filtering by target/driver).
+# However, 'check' tests are limited to single functions that take no arguments
+# and return no results, while these tests are specifically interested in
+# testing such structural properties of programs.
```

Comment:
@benvanik / @pzread / others ... could I get some initial feedback (bikeshedding welcome) on this approach to increasing test coverage? I can figure out the CI configuration and populate more tests if this approach seems good.

Note that there is some overlap with
* https://github.com/openxla/iree/blob/main/tools/test/multiple_args.mlir
* https://github.com/openxla/iree/blob/main/tools/test/multiple_exported_functions.mlir

(I figure those are tests for the tools, while these could be tests for each backend/runtime)

Reply:
Just curious, is it possible to use or extend `iree_bytecode_module + iree_run_module_test` for the use cases here? Although it might need to write more rules.

I like the cleanliness of the lit tests here as we can clearly describe how the mlir is compiled and run in the same file just with a few lines. Wondering if we can extend `iree_run_module_test` to achieve a similar simplicity.

Another advantage to use `iree_bytecode_module + `ree_run_module_test` is that `iree_bytecode_module` will populate the ASAN/TSAN flags following the CMake configuration.

Example: https://github.com/openxla/iree/blob/77a4a0ce8d945ea90b52d2e70afb33d4cf954f5b/tools/test/CMakeLists.txt#L82

Reply:
Good ideas! After successful compilation, `iree_run_module_test` would indeed be a good fit, though more verbose in the build files. For the case that motivated this though, compilation was failing, and I don't see a good way to make the build step `iree_bytecode_module` support `WILL_FAIL` / `XFAIL`.

Here's an idea:
* Use `iree_bytecode_module` together with `iree_run_module_test` for targets that can compile successfully
* For targets that fail to run correctly, use `XFAIL` in the test rule
* For targets that fail to compile, omit those rules and use lit
* The `iree_bytecode_module` compilation will ignore the lit comments in the file

Reply:
I see. I feel like if we expect some `iree_bytecode_module` to fail, essentially they become ctests and they are host tests (on x86_64 or where `iree-compile` runs). For the mlir that can be compiled successfully, technically they should also be part of the ctests and practically we can test those compilations in the build stage.

So yeah, for mlir that can compile successfully, we can run `iree_bytecode_module` in the build stage and test with `iree_run_module_test` on the target devices. For mlir can't be compiled for now, put them in a ctest suite with XFAIL, so we will know when they can be compiled

Reply:
I have this for the XFAIL test right now:

```
// WebGPU fails at the moment. Flip the return code to success using 'not'
// RUN: not iree-compile --iree-hal-target-backends=webgpu \
//   --iree-stream-resource-alias-mutable-bindings=true %s 2>&1 | \
//   FileCheck %s --check-prefix=WEBGPU-ERRORS

// WEBGPU-ERRORS: Tint reported 1 error(s) for a SPIR-V program, see diagnostics:
// WEBGPU-ERRORS: error: entry point 'd0' references multiple variables that use the same resource binding @group(0), @binding(2)
```

a bit more expressive than just XFAIL on a test, since it checks the error message.

Reply:
I'm leaning more towards lit for these tests than to `iree_run_module_test`. I think build system functions would be quite verbose and add distance between the programs and their inputs/outputs, even with test suite wrappers around the base functions. `iree-run-module` is also synchronous, so I couldn't even leverage it (as-is) for WebGPU correctness testing. I think these tests would be most useful for compilation and runtime correctness on host platforms (Linux/Windows/macOS) where lit is available and they don't necessarily need to run on all platforms (Android, iOS, RISC-V, web, etc.).